docs/proposals/20210112-encryption-roviders.md

```diff
@@ -0,0 +1,112 @@
+    customProviderFile: |
+      apiVersion: apiserver.config.k8s.io/v1
+      kind: EncryptionConfiguration
+      resources:
+      - resources:
+        - secrets
+        providers:
+        - identity: {}
+        - aescbc:
+            keys:
+            - name: key1
+            secret: <BASE 64 ENCODED SECRET>
+```
+
+### pre-flight checks
+ * Cluster is healthy.
+ * Current Encryption Providers state/configuration is valid and identical on all control plane nodes.
+ * Ensure Custom Encryption Providers configuration file is valid (if any). 
+
+### Enable Encryption Providers for new cluster
+* Generate a valid configuration file with the `identity` provider set last.
+* Sync the configuration file to all Control Plane nodes. 
+* Set the required KubeAPI configuration and deploy KubeAPI.
+
+### Enable Encryption Providers for existing cluster
+
+* Ensure there is no Encryption Provider Config (manually added by the user, broken previos enable process, etc..) present.
+* Generate a valid configuration file with the `identity` provider set last.
+* Sync the configuration file to all Control Plane nodes. 
+* Update and restart KubeAPI on all nodes.
+* Rewrite secrets to ensure they are encrypted successfully.
+
+### Disable Encryption Providers for existing cluster 
+
+* Read the current active Encryption Provider configuration from control plane nodes.
+* Mutate the configuration to add `identity` provider first and the active provider last.
+* Sync the configuration file to all Control Plane nodes. 
+* Restart KubeAPI on all control plane nodes.
+* Rewrite secrets to ensure they are decrypted successfully.
+* Update KubeAPI configuration to remove the Encryption Provider configuration and restart KubeAPI on all control plane nodes. 
+* Remove the old configuration file from all control plane nodes.
+
+### Rotate Encryption Provider Key for existing cluster
+
+* Read the current active Encryption Provider configuration from control plane nodes.
+* Generate a new encryption key.
+* Mutate the configuration file to include the new key first, current key second and `identity` last.
+* Sync the updated configuration file to all control plane nodes and restart KubeAPI.
+* Rewrite all secrets to ensure they are encrypted with the new key.
+* Mutate the configuration file again to remove the old key.
+* Sync the updated configuration file to all control plane nodes and restart KubeAPI.
+
+### Apply Custom Encryption Provider file
+This use case is useful for users who would like to utilize an external KMS provider or specify resources other than secrets for encryption. In this case, KubeOne will not manage the content of the file, it will only validate it to make sure it's syntactically valid. Additionally, KubeOne will not rewrite the resources in this case. 
+
+* Ensure the configuration file is valid. 
+* Sync the configuration file to all control plane nodes.
+* Restart KubeAPI on all nodes. 
```

Comment:
Custom Encryption Provider require kubeapi-server to have unix socket mounted inside the kubeapi pod (to let kubeapi communicate over it with KMS). Yet, user has no chance to influence on this, so I'm not really sure if we need this.

Let's not provide possibility to specify custom file. This will give us time to figure out a way to have custom KMS providers without regrets on our API.

Reply:
Good point :+1: